src/Microsoft.Android.Build.BaseTasks/AsyncTask.cs

```diff
@@ -0,0 +1,431 @@
+		readonly ManualResetEvent errorDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent warningDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent customDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent telemetryDataAvailable = new ManualResetEvent(false);
+		readonly ManualResetEvent taskCancelled = new ManualResetEvent(false);
+		readonly ManualResetEvent completed = new ManualResetEvent(false);
+		bool isRunning = true;
+		object eventlock = new object();
+		int uiThreadId = 0;
+
+		/// <summary>
+		/// Indicates if the task will yield the node during tool execution.
+		/// </summary>
+		public bool YieldDuringToolExecution { get; set; }
+
+		/// <summary>
+		/// The cancellation token to notify the cancellation requests
+		/// </summary>
+		public CancellationToken CancellationToken => cts.Token;
+
+		/// <summary>
+		/// Gets the current working directory for the task, which is captured at task 
+		/// creation time from <see cref="Directory.GetCurrentDirectory"/>.
+		/// </summary>
+		protected string WorkingDirectory { get; private set; }
+
+		[Obsolete("Do not use the Log.LogXXXX from within your Async task as it will Lock the Visual Studio UI. Use the this.LogXXXX methods instead.")]
+		private new TaskLoggingHelper Log => base.Log;
+
+		/// <summary>
+		/// Initializes the task.
+		/// </summary>
+		public AsyncTask()
+		{
+			YieldDuringToolExecution = false;
+			WorkingDirectory = Directory.GetCurrentDirectory();
+			uiThreadId = Thread.CurrentThread.ManagedThreadId;
+		}
+
+		public void Cancel() => taskCancelled.Set();
+
+		protected void Complete(System.Threading.Tasks.Task task)
+		{
+			if (task.Exception != null)
+			{
+				var ex = task.Exception.GetBaseException();
+				LogError(ex.Message + Environment.NewLine + ex.StackTrace);
+			}
+			Complete();
+		}
+
+		public void Complete() => completed.Set();
+
+		public void LogDebugTaskItems(string message, string[] items)
+		{
+			LogDebugMessage(message);
+
+			if (items == null)
+				return;
+
+			foreach (var item in items)
+				LogDebugMessage("	{0}", item);
+		}
+
+		public void LogDebugTaskItems(string message, ITaskItem[] items)
+		{
+			LogDebugMessage(message);
+
+			if (items == null)
+				return;
+
+			foreach (var item in items)
+				LogDebugMessage("	{0}", item.ItemSpec);
+		}
+
+		public void LogTelemetry(string eventName, IDictionary<string, string> properties)
+		{
+			if (uiThreadId == Thread.CurrentThread.ManagedThreadId)
+			{
+#pragma warning disable 618
+				Log.LogTelemetry(eventName, properties);
+				return;
+#pragma warning restore 618
+			}
+
+			var data = new TelemetryEventArgs() {
+					EventName = eventName,
+					Properties = properties
+			};
+			EnqueueMessage(telemetryMessageQueue, data, telemetryDataAvailable);
+		}
+
+		public void LogMessage(string message) => LogMessage(message, importance: MessageImportance.Normal);
```

Comment:
This is probably a "me" thing, but this line just looks "too wide"; I'd like it more if it were line wrapped:

```csharp
// Option 1
public void LogMessage (string message) =>
    LogMessage (message, importance: MessageImportance.Normal);

// Option 2
public void LogMessage (string message)
    => LogMessage (message, importance: MessageImportance.Normal);
```

I'm not sure I have a preference between Options (1) or (2), just that I don't like "really wide" lines.